### PR TITLE
Convert single-argument %-style format calls

### LINKED ISSUE #3549: `printf-style-formatting` ignores direct-string interpolation
Given:

```py
"%s" % "foo"
"%s" % ("foo",)
```

Ruff will raise UP031 for the latter, but not the former.

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

Today, we only convert %-style format calls if the right-hand side is a dictionary or tuple literal. However, I _think_ we can support a larger set of expressions on the right-hand side.

There are three cases to consider...

1. The template string contain at least one named value (`"%(foo)s" % x`). In that case, the right-hand side _has_ to resolve to a map, so we treat it as kwargs (e.g., `"{foo}".format(**x)`).
2. The template string contains multiple positional values (`"%s %s" % x`). In that case, the right-hand side _has_ to resolve to a tuple, so we treat it as kwargs (e.g., `"{} {}".format(*x)`).
3. The template string contains a single value (`"%s" % x`). In that case, we can't be certain whether the right-hand side is a single expression, or a tuple containing a single expression -- so unfortunately, we have to continue ignoring those.

There's one case that's not formally covered above, which is a template string that mixes named and positional values, like `"%(foo)s %s"`. But I'm unable to come up with a right-hand side that's valid given mixed named and positional values. Does such a case exist? I believe it's impossible.

Closes #3549.

### CODE REVIEW

**Comment by andersk** [comment]
Case 3 does not hold.

```python
x = (1,)
print("%s" % x)
print("{}".format(x))
```

**Comment by andersk** [comment]
Case 3 does not hold.

```python
x = (1,)
print("%s" % x)
print("{}".format(x))
```

**Comment by charliermarsh** [comment]
Ah that's a shame, I'd guess that's a small minority of cases.

**Comment by charliermarsh** [comment]
I can't think of any way to avoid that, so I guess we have to continue omitting those cases.

**Comment by charliermarsh** [comment]
(Removed, and amended the PR summary.)

File: crates/ruff/src/rules/pyupgrade/rules/printf_string_formatting.rs [hunk 1]
```diff
@@ -351,12 +353,52 @@ pub(crate) fn printf_string_formatting(
...
+            } else {
+                // Otherwise, if we have a single field, we can't make any assumptions about the
+                // right-hand side. It _could_ be a tuple, but it could also be a single value,
+                // and we can't differentiate between them.
+                // For example:
+                // ```python
+                // x = (1,)
+                // print("%s" % x)
+                // print("{}".format(x))
+                // ```
```
Comment (MichaReiser) [inline crates/ruff/src/rules/pyupgrade/rules/printf_string_formatting.rs:398]: I love the comments!

File: crates/ruff/src/rules/pyupgrade/rules/printf_string_formatting.rs [hunk 2]
```diff
@@ -335,7 +335,9 @@ pub(crate) fn printf_string_formatting(
     }
 
     // Parse each string segment.
-    let mut format_strings = vec![];
+    let mut num_positional = 0;
+    let mut num_keyword = 0;
```
Comment (MichaReiser) [inline crates/ruff/src/rules/pyupgrade/rules/printf_string_formatting.rs:339]: Nit: A boolean for keywords seems sufficient
```suggestion
    let mut has_keyword_arguments = false;
```

**Review by MichaReiser — APPROVED** [review]
LGTM. I can't say much about the inference rules.

**Comment by konstin** [comment]
I think the proper way to ensure the cases are correctly covered would be reading the cpython source code, but i couldn't find the percent formatting source code, only the tests: https://github.com/python/cpython/blob/5e6661bce968173fa45b74fa2111098645ff609c/Lib/test/test_format.py

**Review by konstin — APPROVED** [review]

File: crates/ruff/src/rules/pyupgrade/rules/printf_string_formatting.rs (hunk 2 above)
Comment (charliermarsh) [reply crates/ruff/src/rules/pyupgrade/rules/printf_string_formatting.rs]: Ah, we need to know if there's _more than_ one positional argument. If we have exactly one positional argument, we can't fix it unfortunately :sob:

**Pull request merged**